Add demo instance page (#124)
SubstraFoundation/substra-documentation: src/setup/demo.md

# Hands on Substra demo

## Let us know

TODO: fr/en page

Check this [page](https://www.substra.ai/fr/demo) and we will then provide you with credentials!

## 1. Configuration

### Install the Substra client

> Substra is not yet compatible with windows, still you can manage to make it work the Windows Linux Subsystem, see the [documentation](https://doc.substra.ai/usage.html?highlight=wsl#installation) or inside a Linux virtual machine.

It will install the CLI (interactive command line interface) & the Python SDK (ready to be used in your Python scripts).

```sh
pip install substra
# Or if a specific version is required by the team
pip install substra==0.6.0 --force-reinstall

# Check that Substra is working
substra --version

> ```

Reviewer: Maybe you can just add an exemple of the expected result here

You will also need to install the substratools package to help yousubmit your algo to the Substra network:

```sh
pip install substratools
```

### Edit your hosts file

> As the demo instance is a demo instance, you will have create a link between the server ip address (provided with your credentials) and the 2 nodes domains used in the demo:

Reviewer: ```suggestion
As the demo instance is a demo instance, you will have to create a link between the server ip address (provided with your credentials) and the 2 nodes domains used in the demo:
```

```sh
# Replace <CHANGE_ME> with the ip provided
echo "<CHANGE_ME> substra-backend.node-1.com substra-frontend.node-1.com substra-backend.node-2.com substra-frontend.node-2.com" | sudo tee -a /etc/hosts
```

This will add a line like the following to the file `/etc/hosts`:

```sh
# 01.02.003.004     substra-backend.node-1.com substra-backend.node-2.com substra-frontend.node-1.com substra-frontend.node-2.com
```

In order to check if your local substra client can reach the demo server, you can use the following command:

```sh
# Quick test on backend node-1
curl substra-backend.node-1.com/readiness
```

You should receive an `"OK"` answer, if not please check again your `/etc/hosts` file.

### How to

```sh
# Configure your local substra package with the credentials you received
substra config --profile <PROFILE> -u <USER> -p "<PASSWORD>" http://substra-backend.node-1.com
# Login
substra login --profile <PROFILE>

# Then you are ready to go!

> substra list node --profile <PROFILE>

Reviewer: maybe you could explain what is this command line and the definition of a node

substra list dataset --profile <PROFILE>

> substra list node --profile <PROFILE>

Reviewer: Duplicate from line one

```

## 2. Work with Substra

> **Note**: Each resource is referenced by its hash in Substra, a unique identifier, so if you want to try out the ready to use Titanic example, you will need to edit, at least, the python files and the `description.md` for your assets.

Reviewer: Does it mean change the name of the asset?

### Data preparation

> 

Reviewer: ```suggestion
The first step is partitioning your data, and create a test dataset and a train dataset.
```

Following the Titanic example, you will first need to generate data samples (in `test_data_samples` and `train_data_samples` folders) by updating the `generate_data_samples` script with the path to you dataset and running its:

```sh
pip install -r scripts/requirements.txt
python scripts/generate_data_samples.py
```

### Add dataset & objective

You will now have to prepare the `opener.py` & `metrics.py` files, including their corresponding `description.md` and `Dockerfile` (where you have to add your depedencies). All assets are based on [substratools](https://github.com/SubstraFoundation/substra-tools/tree/master/substratools) classes to ease the process.

The `add_dataset_objective.py` script is in charge of registering data to the Substra network.

> Before running running the script, please be sure you have updated:

Reviewer: ```suggestion
Before running the script, please be sure you have updated:
```

- `profile` in `client = substra.Client(profile_name="<PROFILE>")`
- `DATASET`: `name` & `permissions` (demo nodes are named `MyOrg1MSP` & `MyOrg2MSP`)
- `OBJECTIVE`: `name` & `permissions`
- any different path in regard to project file structure

> You can then run the run:

Reviewer: ```suggestion
You can then run the command:
```

```sh
python3 add_dataset_objective.py
```

### Add your algorithm

You will need to update the `algo.py` file as well as the `description.md` & `Dockerfile`, and finally push it to the platform with the dedicated script:

```sh
python3 add_train_algo.py
```

> In the `algo.py` file, please be sure you have updated:

Reviewer: ```suggestion
In the `algo.py` file, please make sure you have updated:
```

- `profile` in `client = substra.Client(profile_name="<PROFILE>")`
- `ALGO`: `name` &  `permissions`

If you need to test that your algorithm is working without having access to the data, you can use the CLI method `substra run-local` that will use the `fake_data` method from the `opener.py`:

```sh
substra run-local assets/algo \
  --train-opener=assets/dataset/opener.py \
  --test-opener=assets/dataset/opener.py \
  --metrics=assets/objective/ \
  --fake-data-samples
```

### Training & Testing tasks

> ```sh

Reviewer: I think in this part you need to describe precisely what the commands do

  --input-predictions-path assets/pred-test.csv \
  --output-perf-path assets/perf-test.json \
  --log-path assets/logs/test_metrics.log
```

You will need to update or follow the relevant path to add your assets or get the logs and model.

## 3. Resources

### Find help

Feeling lost? Remember, you can use `substra --help` anytime:

```sh
Usage: substra [OPTIONS] COMMAND [ARGS]...

  Substra Command Line Interface.

  For help using this tool, please open an issue on the Github
  repository: https://github.com/SubstraFoundation/substra

Options:
  --version  Show the version and exit.
  --help     Show this message and exit.

Commands:
  add          Add new asset to Substra platform.
  cancel       Cancel execution of an asset.
  config       Add profile to config file.
  describe     Display asset description.
  download     Download asset implementation.
  get          Get asset definition.
  leaderboard  Display objective leaderboard
  list         List assets.
  login        Login to the Substra platform.
  run-local    Run local.
  update       Update asset.
```

You can also get help for specific commands, for example `substra download --help`.

### Get inspired

> The [Titanic example](https://github.com/SubstraFoundation/substra/blob/master/examples/titanic/README.md), included in the Substra repository, is ready to run but you can also have a look at the [community examples](https://github.com/SubstraFoundation/substra-examples):

Reviewer: maybe we could start with that: 
- First: discover the Titanic exemple to start
- Then: add your own asset.

Author: Reorganized as:

> 

Reviewer: should we do a part about the CLI?

Author: What do you mean?
There are examples of CLI usage in this page, in the titanic example page, a (short) section about `substra --help` just before this one, and the link to the CLI documentation just after this section.
Do you think about something in particular that is missing or not clear?

Reviewer: Explain how to connect and what you can see (or link to the CLI documentation page)

# Hands on Substra demo

Check this [page](https://www.substra.ai/en/demo) and get in touch with us, we will then provide you with instructions and credentials!

## 1. Configuration

### Install the Substra client

> Substra is not yet compatible with windows, still you can manage to make it work the Windows Linux Subsystem, see the [documentation](https://doc.substra.ai/usage.html?highlight=wsl#installation) or inside a Linux virtual machine.

> It will install the [CLI](https://github.com/SubstraFoundation/substra/blob/master/references/cli.md#summary) (interactive command line interface) & the [Python SDK](https://github.com/SubstraFoundation/substra/blob/master/references/sdk.md#substrasdk) (ready to be used in your Python scripts). You can also clone the [Substra repository](https://github.com/SubstraFoundation/substra) to get the full examples located in the `example` folder.

Author: Okay!

### Login

You can now configure your local Substra client, *with the credentials you received from us*, to connect to the `node-1` which hosts the organisation `org-1`:

```sh
# Configure the client, with a profile, a user, a password and an url
substra config --profile <PROFILE> -u <USER> -p "<PASSWORD>" http://substra-backend.node-1.com

# Login with the profile you created (located in your home folder: ~/.substra)
substra login --profile <PROFILE>

# You can then list the network nodes that are on the Substra network
substra list node --profile <PROFILE>

# And list, for example, the registered datasets in the node you are logged in
substra list dataset --profile <PROFILE>
```

## 2. Follow Titanic example

You will find everything you need to run a first end to end example with the tabular data of the Titanic: [repository](https://github.com/SubstraFoundation/substra/blob/master/examples/titanic/README.md).

> Please note that each resource is referenced by its hash in Substra, a unique identifier, so if you want to try out the ready to use Titanic example, you will need to edit, at least, the python files and the `description.md` for the assets (like adding a `print()` statement and editing the `description.md` files).

Reviewer 2: I don't understand this part. Is this optional? What would be the goal of adding a `print()` statement and where exactly? This part is really confusing.

Author: I should not be required with the `exist_ok` option, I wanted to confirm this but I can get any running instance...

## 3. Work with your own assets

> **Note**: This is demo instance, please do not use any sensitive asset on this platform!

The special [debugging section](https://github.com/SubstraFoundation/substra/tree/master/examples/debugging) can be off help preparing your asset, please have a look!

### Data preparation

The first step is partitioning your data, and create a test dataset and a train dataset. Following the Titanic example, you will first need to generate data samples (in `test_data_samples` and `train_data_samples` folders) by updating the `generate_data_samples` script with the path to you dataset and running it:

```sh
pip install -r scripts/requirements.txt
python scripts/generate_data_samples.py
```

### Add dataset & objective

You will now have to prepare the `opener.py` & `metrics.py` files, including their corresponding `description.md` and `Dockerfile` (where you have to add your depedencies). All assets are based on [substratools](https://github.com/SubstraFoundation/substra-tools/tree/master/substratools) classes to ease the process.

> The `add_dataset_objective.py` script is in charge of registering data to the `organisation`.

Reviewer 2: You mean once you have finished preparing the opener and metrics files? Or independently and in parallel?

Author: It's in the order of the text: 
1. prepare the opener and metrics
2. edit the required elements in the script & run it

Reviewer 2: OK, so for total clarity I would do:

```suggestion
The `add_dataset_objective.py` script is in charge of registering your dataset as an asset of the `organisation` you belong to. With your dataset will also be registered the `opener.py` & `metrics.py` files.
```

Before running the script, please make sure you have updated:

- `profile` in `client = substra.Client(profile_name="<PROFILE>")`, using the profile you created before
- `DATASET`: `name` & `permissions` (demo organisations are named `MyOrg1MSP` & `MyOrg2MSP`)
- `OBJECTIVE`: `name` & `permissions` (demo organisations are named `MyOrg1MSP` & `MyOrg2MSP`)

> - any different path in regard to project file structure

Reviewer 2: All this is to be verified in `add_dataset_objective.py`? If yes please add it for clarity/disambiguation

Author: Not sure to understand @bowni, as stated, "Before running the script, please make sure you have updated:" + elements to edit.
The section is composed of 3 paragraphs: 
- edit open and metrics (+associated files)
- edit the `add_dataset_objective.py` script
- run the script

Please let me know how I could make it clearer

Reviewer 2: Like that:

```suggestion
Before running the `add_dataset_objective.py` script, please make sure you edit it to update the following:

- `profile` in `client = substra.Client(profile_name="<PROFILE>")`, using the profile you created before
- `DATASET`: `name` & `permissions` (demo organisations are named `MyOrg1MSP` & `MyOrg2MSP`)
- `OBJECTIVE`: `name` & `permissions` (demo organisations are named `MyOrg1MSP` & `MyOrg2MSP`)
- any different path in regard to project file structure
```

You can then run the command:

```sh
python3 add_dataset_objective.py

> ```

Reviewer 2: Cool - what is the confirmation message that it ran correctly? It is reassuring to be able to see it here

Author: done!

### Add your algorithm

You will need to update the `algo.py` file as well as the `description.md` & `Dockerfile`, and finally push it to the platform with the dedicated script:

```sh
python3 add_train_algo.py

> ```

Reviewer 2: Same here, what is the expected return?

Author: done!

# Predict
python assets/algo/algo.py predict \
  --debug \
  --opener-path assets/dataset/opener.py \
  --data-samples-path assets/train_data_samples \
  --output-predictions-path assets/pred-train.csv \
  --models-path assets/model/ \
  --log-path assets/logs/train_predict.log \
  model

# Performance
python assets/objective/metrics.py \
  --debug \
  --opener-path assets/dataset/opener.py \
  --data-samples-path assets/train_data_samples \
  --input-predictions-path assets/pred-train.csv \
  --output-perf-path assets/perf-train.json \
  --log-path assets/logs/train_metrics.log

# Test / predict
python assets/algo/algo.py predict \
  --debug \
  --opener-path assets/dataset/opener.py \
  --data-samples-path assets/test_data_samples \
  --output-predictions-path assets/pred-test.csv \
  --models-path assets/model/ \
  --log-path assets/logs/test_predict.log \
  model

# Test / performance
python assets/objective/metrics.py \
  --debug \
  --opener-path assets/dataset/opener.py \
  --data-samples-path assets/test_data_samples \
  --input-predictions-path assets/pred-test.csv \
  --output-perf-path assets/perf-test.json \
  --log-path assets/logs/test_metrics.log

> ```

Reviewer 2: Ok - so I guess from this that:

- we start with training a model
- then we launch a predict task, which seems to be needed to be able to launch the computation of the performance metric - is that it?
- on the test set, same thing: predict + performance.

If this is correct, could you please group this in 2 groups: 1 for train, 1 for test? It will be clearer.

Author: Okay!

# Hands-on Substra demo

Check this [page](https://www.substra.ai/en/demo) and get in touch with us, we will then provide you with instructions and credentials!

## 0. Pre-requisites & Overview of the demo

### Terminology

> To follow this tutorial for using the demo instance, it is assumed you have read the documentation of the Substra Framework and are familiar with its core concepts. In particular the following terms: organisation, asset, algo, objective, metric, opener, train task, test task.

Reviewer: Seems a little bit heavy to must read "all" documentation. but why not.

Author: Indeed, cc @bowni 
I'll add the link to the glossary here, by the way

### Technical pre-requisites

The demo instance is deployed on a cloud infrastructure. To use it you will need:

- the credentials that were provided to you after you contacted us for accessing the demo instance
- access to a computer (e.g. a laptop, a VM...) with an internet connection, on which you have sudo privileges. Preferably not a Windows machine, although this is still manageable (see below).

### Overview: what you can do with the demo instance

Although this can be inferred from the below tutorial by experienced readers, it might be good to be explicit about the proposed demo:

> - Substra Framework is meant to be instantiated as a distributed application among multiple partners. However, for this simple demo to be be accessible without any deployment operation and allow you to focus on the data scientist perspective, it is installed as a single instance on a cloud VM, in which 2 `organisations` have been created and live independently.

Reviewer: ```suggestion
- Substra Framework is meant to be instantiated as a distributed application among multiple partners. However, for this simple demo to be accessible without any deployment operation and allow you to focus on the data scientist perspective, it is installed as a single instance on a cloud VM, in which 2 `organisations` have been created and live independently.
```

> - The credentials provided to you grant you access to only one of these 2 `organisations`. You will not be able to access or tamper with any asset of the second one; you can consider it as another company to which you don't have access.

Reviewer: You have only access to one organization? so you run everything on one organization if you want to add your asset? is it relevant?

Author: No, 2 orgs are instanciated and you can use the second as well

Author: According to what people want to do, we can create credentials to both orgs

### Login

You can now configure your local Substra client, *with the credentials that were provided to access the demo instance*, to connect to the `node-1` which hosts the organisation `org-1`:

```sh
# Configure the client, with a profile, a user, a password and an url
substra config --profile <PROFILE> -u <USER> -p "<PASSWORD>" http://substra-backend.node-1.com

# Login with the profile you created (located in your home folder: ~/.substra)
substra login --profile <PROFILE>

# You can then list the network nodes that are on the Substra network
substra list node --profile <PROFILE>

# And list, for example, the registered datasets in the node you are logged in
substra list dataset --profile <PROFILE>
```

## 2. Follow Titanic example

With this example you will find everything you need to run a first end-to-end example with the tabular data of the Titanic: see this [repository](https://github.com/SubstraFoundation/substra/blob/master/examples/titanic/README.md) for further details.

If you feel more comfortable working with local files, you can clone the [Substra repository](https://github.com/SubstraFoundation/substra) to get the full examples located in the `/example` folder.

> Please note that each resource is referenced by its hash in Substra, a unique identifier, so if you want to try out the ready-to-use Titanic example, you will need to edit, at least, the python files and the `description.md` for the assets (like adding a `print()` statement and editing the `description.md` files).

Reviewer: Is it working if everybody add a "print()"? A random name is not better?

Author: In fact I was thinking about this section, but you are right. But printing something won't break anything
